x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.http.retry;
+
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.RetryableAction;
+import org.elasticsearch.inference.InferenceResults;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.inference.external.http.HttpResult;
+import org.elasticsearch.xpack.inference.external.http.sender.Sender;
+import org.elasticsearch.xpack.inference.logging.ThrottlerManager;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.InferencePlugin.UTILITY_THREAD_POOL_NAME;
+
+public class RetryingHttpSender implements Retrier {
+    private final Sender sender;
+    private final ThrottlerManager throttlerManager;
+    private final Logger logger;
+    private final RetrySettings retrySettings;
+    private final ThreadPool threadPool;
+    private final Executor executor;
+
+    public RetryingHttpSender(
+        Sender sender,
+        ThrottlerManager throttlerManager,
+        Logger logger,
+        RetrySettings retrySettings,
+        ThreadPool threadPool
+    ) {
+        this(sender, throttlerManager, logger, retrySettings, threadPool, threadPool.executor(UTILITY_THREAD_POOL_NAME));
+    }
+
+    // For testing only
+    RetryingHttpSender(
+        Sender sender,
+        ThrottlerManager throttlerManager,
+        Logger logger,
+        RetrySettings retrySettings,
+        ThreadPool threadPool,
+        Executor executor
+    ) {
+        this.sender = Objects.requireNonNull(sender);
+        this.throttlerManager = Objects.requireNonNull(throttlerManager);
+        this.logger = Objects.requireNonNull(logger);
+        this.retrySettings = Objects.requireNonNull(retrySettings);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.executor = Objects.requireNonNull(executor);
+    }
+
+    private class InternalRetrier extends RetryableAction<InferenceResults> {
```

Comment:
Leveraging the `RetryableAction` made the code much simpler and gets exponential backoff for free